src/Npgsql/NpgsqlTypes/NpgsqlCube.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+#pragma warning disable 1591
+
+// ReSharper disable once CheckNamespace
+namespace NpgsqlTypes
+{
+    public readonly struct NpgsqlCube : IEquatable<NpgsqlCube>
+    {
+        public const int MaxDimensions = 100;
+        
+        internal static readonly Regex Regex =
+            new(@"^(\[(?=\(.+\)\,\())?((\()?(-?\d+\.?\d*?(?:,-?\d+(?:\.?\d*))*)(?(3)\)|)(,\((-?\d+\.?\d*(?:,-?\d+\.?\d*)*)\))?)(?(1)\]|)$");
+        
+        /// <summary>
+        /// The lower left coordinates of the cube.
+        /// </summary>
+        public List<double> LowerLeft { get; }
+        
+        /// <summary>
+        /// The upper right coordinates of the cube.
+        /// </summary>
+        public List<double> UpperRight { get; }
+
+        /// <summary>
+        /// Constructs a point from coordinates.
+        /// </summary>
+        /// <param name="coords">The coordinates of the cube.</param>
+        public NpgsqlCube(IEnumerable<double> coords) : this(coords.ToArray()) { }
+        
+        /// <summary>
+        /// Constructs a point from coordinates.
+        /// </summary>
+        /// <param name="coords">The coordinates of the cube.</param>
+        public NpgsqlCube(params double[] coords) : this(coords, coords) { }
+
+        /// <summary>
+        /// Constructs a cube from coordinates.
+        /// </summary>
+        /// <param name="lowerLeft">The lower-left coordinates.</param>
+        /// <param name="upperRight">The upper-right coordinates.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown if the number of dimensions in the lower-left and upper-right coordinates are different.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if the cube has too many dimensions.</exception>
+        public NpgsqlCube(IEnumerable<double> lowerLeft, IEnumerable<double> upperRight)
+        {
+            LowerLeft = lowerLeft.ToList();
+            UpperRight = upperRight.ToList();
+
+            if (UpperRight.Count != LowerLeft.Count)
+                throw new ArgumentException("Not a valid cube: Different point dimensions in " + ToString(), nameof(upperRight));
```

Comment:
Since we fill missing coordinates with zero, we probably don't even need to do this check in the constructor, the data will end up normalized when written to the database.